ICSharpCode.Decompiler/IL/Transforms/NullCoalescingTransform.cs

```diff
@@ -83,7 +87,173 @@ bool TransformRefTypes(Block block, int pos, StatementTransformContext context)
 				ILInlining.InlineOneIfPossible(block, pos, InliningOptions.None, context);
 				return true;
 			}
+			// stloc obj(valueInst)
+			// if (comp(ldloc obj == ldnull)) {
+			//		throw(...)
+			// }
+			// =>
+			// stloc obj(if.notnull(valueInst, throw(...)))
+			if (context.Settings.ThrowExpressions && trueInst is Throw throwInst) {
+				context.Step("NullCoalescingTransform (reference types + throw expression)", stloc);
+				throwInst.resultType = StackType.O;
+				stloc.Value = new NullCoalescingInstruction(NullCoalescingKind.Ref, stloc.Value, throwInst);
+				block.Instructions.RemoveAt(pos + 1); // remove if instruction
+				ILInlining.InlineOneIfPossible(block, pos, InliningOptions.None, context);
+				return true;
+			}
 			return false;
 		}
+
+		delegate bool PatternMatcher(ILInstruction input, out ILInstruction output);
+
+		/// <summary>
+		/// stloc v(value)
+		/// if (logic.not(call get_HasValue(ldloca v))) throw(...)
+		/// ... Call(arg1, arg2, call GetValueOrDefault(ldloca v), arg4) ...
+		/// =>
+		/// ... Call(arg1, arg2, if.notnull(value, throw(...)), arg4) ...
+		/// 
+		/// -or-
+		/// 
+		/// stloc v(value)
+		/// stloc s(ldloca v)
+		/// if (logic.not(call get_HasValue(ldloc s))) throw(...)
+		/// ... Call(arg1, arg2, call GetValueOrDefault(ldloc s), arg4) ...
+		/// =>
+		/// ... Call(arg1, arg2, if.notnull(value, throw(...)), arg4) ...
+		/// </summary>
+		bool TransformThrowExpressionValueTypes(Block block, int pos, StatementTransformContext context)
+		{
+			if (pos + 2 >= block.Instructions.Count)
+				return false;
+			if (!(block.Instructions[pos] is StLoc stloc))
+				return false;
+			int offset = 1;
+			ILVariable v = stloc.Variable;
+			// alternative pattern using stack slot containing address of v
+			if (block.Instructions[pos + offset] is StLoc addrCopyStore
+				&& addrCopyStore.Variable.Kind == VariableKind.StackSlot
+				&& addrCopyStore.Value.MatchLdLoca(v)
+				&& pos + 3 < block.Instructions.Count)
+			{
+				offset++;
+				// v turns into s in the pattern above.
+				v = addrCopyStore.Variable;
+				if (!(v.StoreCount == 1 && v.LoadCount == 2 && v.AddressCount == 0))
+					return false;
+			} else {
+				if (!(v.StoreCount == 1 && v.LoadCount == 0 && v.AddressCount == 2))
+					return false;
+			}
+			if (!block.Instructions[pos + offset].MatchIfInstruction(out var condition, out var trueInst))
+				return false;
+			if (!(Block.Unwrap(trueInst) is Throw throwInst))
+				return false;
+			if (!condition.MatchLogicNot(out var arg))
+				return false;
+			if (!MatchNullableCall(arg, NullableLiftingTransform.MatchHasValueCall))
+				return false;
+			var throwInstParent = throwInst.Parent;
+			var throwInstChildIndex = throwInst.ChildIndex;
+			var nullCoalescingWithThrow = new NullCoalescingInstruction(
+				NullCoalescingKind.NullableWithValueFallback,
+				stloc.Value,
+				throwInst);
+			var resultType = NullableType.GetUnderlyingType(v.Type).GetStackType();
```

Comment:
In the case of `stloc s(ldloca v)`, doesn't this always result in `resultType == Ref`?